### PR TITLE
Replace Shape elimination rewrite rule with generation-time logic

### PR DESCRIPTION
**Pull request by justinchuby**
Moves the EP-specific Shape-vs-ReduceSum decision from a post-hoc rewrite rule into graph-construction time via `ep_capabilities().supports_shape`.

## Changes

- **New helper `_mask_seq_len()`** in `_common.py`: emits `Shape(attention_mask, start=1, end=2)` on most EPs, and `ReduceSum+ReduceMax` on WebGPU where Shape cannot run at runtime
- **Replace both call sites** in `create_attention_bias()` and `create_padding_mask()`
- **Delete `_eliminate_shape.py`** (−221 lines) and its tests (−172 lines)
- **Remove `EliminateShape` stage** from `_optimizations.py`

## Motivation

Components should be EP-aware at construction time, not rely on cleanup rewrites. The `EpCapabilities.supports_shape` field already existed (WebGPU: `False`). This refactor closes the gap: instead of generating Shape nodes and then rewriting them away, we never emit them on WebGPU in the first place.

**Net: +29/−407 lines.**

## Testing

All 2282 tests pass. Lintrunner clean.

### CODE REVIEW

File: src/mobius/_optimizations.py [hunk 1]
```diff
@@ -286,9 +285,6 @@ def _get_optimization_passes(
         lower.append(("SeparateRoPE", list(separate_rope_rules())))
         lower.append(("UnpackQKV", list(unpack_qkv_rules())))
 
-    if not caps.supports_shape:
-        lower.append(("EliminateShape", list(eliminate_shape_rules())))
-
     return fuse, lower
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/_optimizations.py:288]: Fixed in commit 50b0ba2: both RotaryAttentionToGQA and AttentionToGQA now derive total_seq_len via ReduceMax(reduce_sum, axis=0) — no Shape op emitted from either GQA rewrite rule.

File: src/mobius/_optimizations.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
 2. **Fusion** — promote standard ops to EP-supported fused ops
    (GQA, SkipNorm, GeluFusion). Gated by ``(ep, dtype)`` and ``model_role``.
 3. **Lowering** — decompose ops the EP cannot execute
-   (SeparateRoPE, EliminateShape).
+   (SeparateRoPE).
 4. **Fold** — final dead-node removal and constant folding.
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/_optimizations.py:15]: Fixed in commit 50b0ba2: optimize_model() docstring now reads 'SeparateRoPE for DML. Note: WebGPU Shape avoidance is handled at graph-construction time via _mask_seq_len()' — consistent with the module-level docstring.

File: src/mobius/components/_common.py [hunk 3]
```diff
@@ -158,6 +160,27 @@ def forward(self, op: builder.OpBuilder, x: ir.Value):
...
+    """Extract the sequence length (dim 1) from attention_mask.
+
+    On EPs that support Shape (all except WebGPU), this emits
+    ``Shape(attention_mask, start=1, end=2)`` — a single op.
+
+    On WebGPU (``supports_shape=False``), Shape cannot run at runtime, so we
+    use a data-driven equivalent: ``ReduceMax(ReduceSum(attention_mask, axis=1))``
+    which sums each row (number of valid tokens) and takes the max across the
+    batch to get the total sequence length.
+    """
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/components/_common.py:173]: Fixed in commit 85b4e3b: added TestMaskSeqLen to src/mobius/components/_common_test.py — test_supports_shape_true_emits_shape_op (default EP: Shape emitted, no ReduceSum/ReduceMax) and test_supports_shape_false_emits_reduce_ops_not_shape (WebGPU EP: ReduceSum+ReduceMax, no Shape). Both pass.

**Comment by justinchuby** [comment]
This affects downstream fusion, so probably not a good idea. @gramalingam

**Comment by gramalingam** [comment]
> This affects downstream fusion, so probably not a good idea.

Which fusion? The reality is that the same technical challenges can arise in rewrites or when we use conditional code generation, but it is usually easier to handle them in conditional code generation because we deal with a specific instance (as opposed to the general case in a rewrite-rule). And the issue you are running into is precisely the unstated complication with rewrite-rules: namely dependences between rewrite-rules and the need to order them in some fashion, the fragility of rewrites, etc. In conditional code generation, the code tells us the order in which different conditions affect what is emitted, when there is interaction between different features: and it is easier for a developer to change it as needed.

**Comment by gramalingam** [comment]
I am concerned. If we are unable to express this at generation-time, that is a sign, IMHO, of a flaw in the design of the entire system.

**Comment by justinchuby** [comment]
WebGPU adaptation no longer needed.
